lib/web_ui/lib/src/engine/skwasm/skwasm_impl/surface.dart

```diff
@@ -22,33 +24,55 @@ class SkwasmSurface {
   SkwasmSurface._fromHandle(this._handle);
   final SurfaceHandle _handle;
   OnRenderCallbackHandle _callbackHandle = nullptr;
-  final Map<int, Completer<void>> _pendingRenders = <int, Completer<void>>{};
+  final Map<int, Completer<int>> _pendingCallbacks = <int, Completer<int>>{};
 
   void _initialize() {
     _callbackHandle =
       OnRenderCallbackHandle.fromAddress(
         skwasmInstance.addFunction(
-          _onRender.toJS,
-          'vi'.toJS
+          _callbackHandler.toJS,
+          'vii'.toJS
         ).toDart.toInt()
       );
-    surfaceSetOnRenderCallback(_handle, _callbackHandle);
+    surfaceSetCallbackHandler(_handle, _callbackHandle);
   }
 
   void setSize(int width, int height) =>
     surfaceSetCanvasSize(_handle, width, height);
 
   Future<void> renderPicture(SkwasmPicture picture) {
-    final int renderId = surfaceRenderPicture(_handle, picture.handle);
-    final Completer<void> completer = Completer<void>();
-    _pendingRenders[renderId] = completer;
+    final int callbackId = surfaceRenderPicture(_handle, picture.handle);
+    return _registerCallback(callbackId);
+  }
+
+  Future<ByteData> rasterizeImage(SkwasmImage image, ui.ImageByteFormat format) async {
+    final int callbackId = surfaceRasterizeImage(
+      _handle,
+      image.handle,
+      format.index,
+    );
+    final int context = await _registerCallback(callbackId);
+    final SkDataHandle dataHandle = SkDataHandle.fromAddress(context);
+    final int byteCount = skDataGetSize(dataHandle);
+    final Pointer<Uint8> dataPointer = skDataGetConstPointer(dataHandle).cast<Uint8>();
+    final Uint8List output = Uint8List(byteCount);
+    for (int i = 0; i < byteCount; i++) {
```

Comment:
Is there any equivalent to `TypedArray.set` for copying one array into another in WasmGC? Then we could create a `Uint8List` view into the linear memory and copy it into the GC space as one bulk operation.

Reply:
You want https://github.com/WebAssembly/gc/issues/367, which sadly won't make WasmGC MVP.

Reply:
Oh well...